From 4588616cd9142d21628d39452c344df99084b8ba Mon Sep 17 00:00:00 2001
From: Razvan A Lupusoru <razvan.a.lupusoru@intel.com>
Date: Thu, 5 Dec 2013 10:42:28 -0800
Subject: Dalvik: Vectorization may dereference expression without null check

BZ: 157035

Klocwork identified issue where dereference may happen even though lhs
and rhs of expression may be null. This patch ensures that dereference
only happens after null check.

Change-Id: Ied582e604d0d411b7f32f59cf13d07b66289e4d9
Category: device enablement
Domain: AOSP.Dalvik-Compiler-ME
Origin: internal
Upstream-Candidate: no, need rework
Signed-off-by: Razvan A Lupusoru <razvan.a.lupusoru@intel.com>
---
 vm/compiler/Vectorization.cpp |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/vm/compiler/Vectorization.cpp b/vm/compiler/Vectorization.cpp
index 08c7ff7..5e81ec5 100644
--- a/vm/compiler/Vectorization.cpp
+++ b/vm/compiler/Vectorization.cpp
@@ -1310,12 +1310,12 @@ static bool haveSafeAccumulationHelper (CompilationUnit *cUnit, BytecodeExpressi
     int lhsVR = -1;
     int rhsVR = -1;
 
-    if (lhs->isVirtualRegister() == true)
+    if (lhs != 0 && lhs->isVirtualRegister() == true)
     {
         lhsVR = dvmExtractSSARegister (cUnit, static_cast<VirtualRegister *>(lhs)->getLowSSAReg());
     }
 
-    if (rhs->isVirtualRegister() == true)
+    if (rhs != 0 && rhs->isVirtualRegister() == true)
     {
         rhsVR = dvmExtractSSARegister (cUnit, static_cast<VirtualRegister *>(rhs)->getLowSSAReg());
     }
-- 
1.7.4.1

